### PR TITLE
Convert tags to topic

### PR DESCRIPTION
**Pull request by chuenlok**
1. refracting the script
2. update structure rules to use cached count
3. add featured topics in browse topic page

### CODE REVIEW

File: script/convert_tags_to_topic.rb [hunk 1]
```diff
@@ -5,34 +5,45 @@ def put_things_in_tag_to_correct_position cards, skip_year
...
-      year.save!
-    end
+    company, topic, year = get_subcards card
+    fill_subcards card, company, topic, year, skip_year
+    updated_cards [company, topic, year]
     i += 1
-    Card.cache.reset_soft if i % 10 == 0
+    Card.cache.reset if i % 10 == 0
+  end
+end
```
Comment (ethn) [inline script/convert_tags_to_topic.rb:14]: such well modularized code!!!

File: script/convert_tags_to_topic.rb [hunk 2]
```diff
@@ -5,34 +5,45 @@ def put_things_in_tag_to_correct_position cards, skip_year
...
+    end
   end
 end
 
+def get_subcards card
+  company = card.fetch trait: :wikirate_company, new: {}
+  topic = card.fetch trait: :wikirate_topic, new: {}
+  year = card.fetch trait: :year, new: {}
+  [company, topic, year]
+end
```
Comment (ethn) [inline script/convert_tags_to_topic.rb:31]: could also do:

```
[:wikirate_company, :wikirate_topic, :year].map do |trait|
  card.fetch trait: trait, new: {}
end
```

(obviously not important here)

File: script/convert_tags_to_topic.rb [hunk 3]
```diff
@@ -5,34 +5,45 @@ def put_things_in_tag_to_correct_position cards, skip_year
...
+
+def update_card card
+  return unless card.changed? && card.item_names.size > 0
+  puts "\tUpdating #{card.name} to #{card.content}".green
+  card.save!
+end
+
 Card::Auth.as_bot do
+  Card::Mailer.perform_deliveries = false
```
Comment (ethn) [inline script/convert_tags_to_topic.rb:46]: we really need to make it default to this in non-web environments.

File: script/convert_tags_to_topic.rb [hunk 4]
```diff
@@ -54,4 +65,5 @@ def put_things_in_tag_to_correct_position cards, skip_year
   puts "#{source_cards_with_tag.size} note+tag cards found. Start Praying".green
   put_things_in_tag_to_correct_position source_cards_with_tag, true
   puts 'Finished source+tag updates!'.green
+  Card::Mailer.perform_deliveries = true
```
Comment (ethn) [inline script/convert_tags_to_topic.rb:68]: why is this needed?

File: script/convert_tags_to_topic.rb [hunk 5]
```diff
@@ -40,7 +51,7 @@ def put_things_in_tag_to_correct_position cards, skip_year
     card.type_id = Card::WikirateTopicID
     puts "Updating #{card.name}'s type to topic".green
     card.save!
-    Card.cache.reset_soft
+    Card.cache.reset
```
Comment (ethn) [inline script/convert_tags_to_topic.rb:54]: interested to hear why you did this.  I don't want people to have to think about this kind of thing, but is it necessary here?

**Comment by ethn** [comment]
this looks good to me. Assuming you've tested the browse pages, please merge and run.

File: script/convert_tags_to_topic.rb (hunk 4 above)
Comment (chuenlok) [inline script/convert_tags_to_topic.rb:68]: Is the `Card::Mailer.perform_deliveries` a global setting? I though I need to re-enable it after the script is done.

File: script/convert_tags_to_topic.rb (hunk 5 above)
Comment (chuenlok) [inline script/convert_tags_to_topic.rb:54]: I remember you asked me to add this when we were desperate with the performance.
